interpreter.go

```diff
@@ -503,15 +503,83 @@ func unparseNumber(v float64) string {
 	return fmt.Sprintf("%.17g", v)
 }
 
-// TODO(sbarzowski) Perhaps it should be a builtin?
-// TODO(sbarzowski) Perhaps we should separate recursive evaluation from serialization?
-// 					Strictly evaluating something may be useful by itself.
-func (i *interpreter) manifestJSON(trace *TraceElement, v value, multiline bool, indent string, buf *bytes.Buffer) error {
-	// TODO(dcunnin): All the other types...
+// manifestJSON converts to standard JSON representation as in "encoding/json" package
+func (i *interpreter) manifestJSON(trace *TraceElement, v value) (interface{}, error) {
 	e := &evaluator{i: i, trace: trace}
 	switch v := v.(type) {
+
+	case *valueBoolean:
+		return v.value, nil
+
+	case *valueFunction:
+		return nil, makeRuntimeError("Couldn't manifest function in JSON output.", i.getCurrentStackTrace(trace))
+
+	case *valueNumber:
+		return v.value, nil
+
+	case *valueString:
+		return v.getString(), nil
+
+	case *valueNull:
+		return nil, nil
+
 	case *valueArray:
-		if len(v.elements) == 0 {
+		result := make([]interface{}, 0, len(v.elements))
+		for _, th := range v.elements {
+			elVal, err := th.getValue(i, trace) // TODO(sbarzowski) perhaps manifestJSON should just take potentialValue
+			if err != nil {
+				return nil, err
+			}
+			elem, err := i.manifestJSON(trace, elVal)
+			if err != nil {
+				return nil, err
+			}
+			result = append(result, elem)
+		}
+		return result, nil
+
+	case valueObject:
+		fieldNames := objectFields(v, true)
+		sort.Strings(fieldNames)
+
+		err := checkAssertions(e, v)
+		if err != nil {
+			return nil, err
+		}
+
+		result := make(map[string]interface{})
+
+		for _, fieldName := range fieldNames {
+			fieldVal, err := v.index(e, fieldName)
+			if err != nil {
+				return nil, err
+			}
+
+			field, err := i.manifestJSON(trace, fieldVal)
+			if err != nil {
+				return nil, err
+			}
+			result[fieldName] = field
+		}
+
+		return result, nil
+
+	default:
+		return nil, makeRuntimeError(
+			fmt.Sprintf("Manifesting this value not implemented yet: %s", reflect.TypeOf(v)),
+			i.getCurrentStackTrace(trace),
+		)
+
+	}
+}
+
+func serializeJSON(v interface{}, multiline bool, indent string, buf *bytes.Buffer) {
```

Comment:
Why can't we rely on `encoding/json` to serialize the interface here?

Reply:
Because I didn't want to change the behavior of multiline/indentation. Having exactly the same output is very valuable for testing against C++ version. Unless "encoding/json" can be convinced to format it in exactly the same way, but I haven't look into that yet.

Reply:
That makes sense